contracts/dual_token/DualStakingToken.sol

```diff
@@ -0,0 +1,362 @@
+                }
+
+                // Process the first service in the set
+                updatedReward = (eligibleServiceRewards[0] * lastAvailableRewards) / totalRewards;
+                updatedTotalRewards += updatedReward;
+                curServiceId = eligibleServiceIds[0];
+                // If the reward adjustment happened to have small leftovers, add it to the first service
+                if (lastAvailableRewards > updatedTotalRewards) {
+                    updatedReward += lastAvailableRewards - updatedTotalRewards;
+                }
+                // Add reward to the overall service reward
+                mapStakerStakerInfos[curServiceId].reward += (updatedReward * rewardRatio) / 1e18;
+                // Set available rewards to zero
+                lastAvailableRewards = 0;
+            } else {
+                // Traverse all the eligible services and add to their rewards
+                for (uint256 i = 0; i < numServices; ++i) {
+                    // Add reward to the service overall reward
+                    curServiceId = eligibleServiceIds[i];
+                    mapStakerStakerInfos[curServiceId].reward += (eligibleServiceRewards[i] * rewardRatio) / 1e18;
+                }
+
+                // Adjust available rewards
+                lastAvailableRewards -= totalRewards;
+            }
+
+            // Update the storage value of available rewards
+            availableRewards = lastAvailableRewards;
+        }
+
+        // Lock activity checker after checkpoint
+        IActivityChecker(activityChecker).lock();
+    }
+
+    /// @dev Re-stakes OLAS service Id as it has been evicted.
+    /// @param serviceId OLAS driven service Id.
+    function restake(uint256 serviceId) external {
+        StakerInfo storage stakerInfo = mapStakerStakerInfos[serviceId];
+        // Check for staker existence
+        if (stakerInfo.account == address(0)) {
+            revert();
+        }
+
+        // Get staking instance address
+        address stakingInstance = stakerInfo.stakingInstance;
+
+        // Perform checkpoint first as there might be more rewards
+        checkpoint(stakingInstance);
+
+        // Get staked service state
+        IStaking.StakingState stakingState = IStaking(stakingInstance).getStakingState(serviceId);
+        // Check for evicted service state
+        if (stakingState != IStaking.StakingState.Evicted) {
+            revert();
+        }
+
+        // Unstake OLAS service
+        IStaking(stakingInstance).unstake(serviceId);
+        // Approve back to staking instance
+        IService(serviceRegistry).approve(stakingInstance, serviceId);
+        // Stake service
+        IStaking(stakingInstance).stake(serviceId);
+    }
+
+    /// @dev Unstakes OLAS service Id and unbonds staking token amount.
+    /// @param serviceId OLAS driven service Id.
+    function unstake(uint256 serviceId) external {
```

Comment:
unstake only from staking instance is prevented by the verifier, correct?

Reply:
Actually not true :) The "owner" of the service to call the `unstake` is the dualStakingToken, so only the real owner of the service Id is able to call the unstake only via dualStakingToken contract interaction.